### PR TITLE
Mapped journal

### PR DESCRIPTION
**Pull request by franz1981**
References:
https://issues.apache.org/jira/browse/ARTEMIS-508

Memory Mapped Implementation of Sequential File + Benchs.
Experimental GCFree Single Writer Journal + Benchs.
Experimental Binary Encoding of Journal Header + Record using aligned sequential stride to perform better on every arch (+ avoid ByteBuffer tricks for write on unaligned addresses).

### CODE REVIEW

**Comment by clebertsuconic** [comment]
Can you do these following commands please?

git remote add upstream <URI for apache github>
git pull --rebase upstream master
git push origin <your-branch-name> -f


there's a merge commit that is not right, and this would fix it.

**Comment by franz1981** [comment]
Thanks Clebert! Just done!

File: artemis-journal/pom.xml [hunk 1]
```diff
@@ -15,57 +15,81 @@
...
+            <version>${project.version}</version>
+        </dependency>
+        <!-- needed to compile the tests -->
+        <dependency>
+            <groupId>junit</groupId>
+            <artifactId>junit</artifactId>
+            <scope>test</scope>
+        </dependency>
+        <dependency>
+            <groupId>net.openhft</groupId>
```
Comment (clebertsuconic) [inline artemis-journal/pom.xml]: what are these? what's the license?

Comment (franz1981) [inline artemis-journal/pom.xml]: These are the licences of the two OpenHFT's dependences:

https://github.com/OpenHFT/Chronicle-Core/blob/master/LICENSE.adoc
https://github.com/OpenHFT/Java-Thread-Affinity/blob/master/LICENSE

And for JMH i've to check:
http://openjdk.java.net/projects/code-tools/jmh/

All of these are the dependences of the tools used to do the benchmark.

I'll follow this:
https://activemq.apache.org/artemis/docs/1.2.0/hacking-guide/code.html

To add the test dependences or there is something more i need to know?the first two licenses are ok?

Comment (franz1981) [inline artemis-journal/pom.xml]: And this is the licence for the JMH tool dependences:

http://hg.openjdk.java.net/code-tools/jmh/file/448bd18c85f6/LICENSE

Comment (mtaylor) [inline artemis-journal/pom.xml]: @franz1981 

The chronicle license is OK as it's Apache.  Could you please add new dependencies to dependency management section of the top level pom with a comment including the license.

The JMH link is LGPL which is not compatible (CatX) with Apache.  The policy we have chosen to adopt for CatX deps in tests is to use a separate profile (usually called "extraTests") that will include and run tests with the CatX dep.  This allows users to do a normal build without pulling down CatX deps.  You can see examples of this in the integration tests.

Comment (franz1981) [inline artemis-journal/pom.xml]: Thanls Martyn,
i've read https://activemq.apache.org/artemis/docs/1.2.0/hacking-guide/code.html
and i've to understand one thing:
that link mentions a "test/pom.xml" as the place where i've to put the OpenHFT's deps (they are indeed test deps) with relative licence + artifact's version while you're suggesting to add it in the main pom.xml (not the test's one), thus where is best place to put them?
And about the "extra-test" the procedure is the same as the normal dependencies or there are other things to be aware of?
If 've get it well, for JMH i've to: 
1) add JMH dep on https://github.com/apache/activemq-artemis/blob/master/tests/extra-tests/pom.xml with artifact's version
2) add JMH on https://github.com/apache/activemq-artemis/blob/master/artemis-journal/pom.xml without artifact's version
Have I missed something?
ASAP i'll fix them with your advices :+1:

Comment (mtaylor) [inline artemis-journal/pom.xml]: Hi @franz1981   

It looks like the tests you've added are unit tests contained within the artemis-journal module.  For this I would add the test deps to the top level pom.xml.  The tests/pom.xml is used for a bunch of integration tests.  I think going forward we should keep everything in the top level pom.xml.  I'll update the hacking guide, my apologies that this wasn't clear.

Here is what I would do:
1. Add JMH dep under the dependencyManagement section of the main pom.xml.
2. Exclude the new tests in the default profile.
3. Create a new profile "extraTests" in the artemis-journal and use this to enable the additional tests and dependencies you've added.

This process is less than ideal, but after some discussion with Apache Legal and other community members we decided this was the best approach to ensure we are always compliant with Apache Licensing.

I'm around on IRC if you want to chat "#activemq"

Thank you

Comment (franz1981) [inline artemis-journal/pom.xml]: Hi Martyn,

apologies more than accepted, i don't want to be so precise, but i've to learn better the project OSS mechanics and I need to ask a lot of things that seem (maybe are) obvious: so i greatly appreciate your patience.
ASAP i'll fix this dep/licence issue; maybe during this night or the next one.
Ok for IRC, i'haven't use it yet, so good to start here :+1:

Comment (franz1981) [inline artemis-journal/pom.xml]: Hi Martyn,

i'm currently trying to fix the deps issue and i've found another problem:
my "tests" are not JUnit tests that could be run like the others, but harness benchmarks, thus in order to not invalidate the results they provides they defined a main method to be run eventually by the dev from the IDE.
They are dev tools to measure something and not something that (in this current form at least) could be embedded in a automatic build chain...
Hence there is no need to create different profiles that exclude these benchs using surefire's plugin, because surefire do not recognize them anyway as proper tests.
Considering these assumptions what do you recommend to do with them?
In this moment i've added them on the main pom with type of license and version (with scope test but only because "bench" doesn't exist in Maven!) and without version and licence type on pom of the journal module (with scope test).

Comment (mtaylor) [inline artemis-journal/pom.xml]: I see.  Yes go ahead and do this, it should be alright.  I can review it once you've updated the pom.

As I am thinking about this more, the underlying issue we originally faced was that we ship the tests as part of the build.  I just checked and we are not shipping the journal tests nor do we depend on them in the integration tests.  So I think we're OK.

Comment (mtaylor) [inline artemis-journal/pom.xml]: There's also this statement in the Apache Legal docs: CAN APACHE PROJECTS RELY ON COMPONENTS UNDER PROHIBITED LICENSES?¶

Apache projects cannot distribute any such components. As with the previous question on platforms, the component can be relied on if the component's licence terms do not affect the Apache product's licensing. For example, using a GPL'ed tool during the build is OK.

Which contradicts my original statement.  We've been a bit overly cautious with introducing new dependencies in case we break license rules.  There are a lot of edge cases that people have brought up in similar discussions so we take a strict rule on any new deps.  But I think you're all good.

Thanks for your patience and for taking the time to contribute!

Comment (franz1981) [inline artemis-journal/pom.xml]: Perfect, i've pushed aand squashed the modification on pom deps...but now Jenkins failed for something that do not seem related to me,but it must be..need some investigation, i do not understand how it's happened

**Comment by franz1981** [comment]
I've squashed all the commits into one and moved all the (micro)benchs to extra-tests.

**Comment by mtaylor** [comment]
@franz1981 Thanks.  +1 from me.  @clebertsuconic you've spent more time reviewing this than I have so I will leave to you to give the final ack.  Cheers

File: artemis-journal/pom.xml [hunk 2]
```diff
@@ -14,7 +14,8 @@
   See the License for the specific language governing permissions and
   limitations under the License.
 -->
-<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/maven-v4_0_0.xsd">
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
```
Comment (clebertsuconic) [inline artemis-journal/pom.xml]: I will revert this change during the merge.
